### PR TITLE
Fee integration

### PR DESCRIPTION
**Pull request by karim-en**
## Description

This PR contains Fee Implementation for Deposit and Withdrawal flows for Ether transfer using Rainbow Bridge (Business requirement).

Specification:
- Deposit and withdrawal can have separate fee percentages.
- By default, the fee collected is locked in the `same contract` that the owner could claim using the function `claim_fee`.
- The fee is transferred to the fee `owner account` instead of the current contract account if the `set_fee_owner` was called.
- All the fee setters are only callable by the owner.
- The `fee_amount` has lower and upper bounds.
- If the the calculated `fee_amount` is higher than the `withdraw/deposit amount` then charge the full amount as a fee and don't transfer anything to the recipient.
- The fee is set in 6 decimal precision `10% = 0.1 * 10e6 = 100_000`
- It is possible to set a particular fee for silos on deposit and withdrawal. The silo is just a NEAR account, so as a side effect, it is possible to set a specific fee to any NEAR account. 

Added public function:
```
fn calculate_fee_amount(&self, amount: U128, fee_type: FeeType, silo: Option<AccountId>) -> U128;
fn set_deposit_fee(&mut self, fee: Option<Fee>);
fn set_withdraw_fee(&mut self, fee: Option<Fee>);
fn set_deposit_fee_per_silo(&mut self, silo: AccountId, fee: Option<Fee>);
fn set_withdraw_fee_per_silo(&mut self, silo: AccountId, fee: Option<Fee>);
fn set_fee_owner(&mut self, owner: Option<AccountId>);
fn claim_fee(&mut self, amount: U128, receiver_id: Option<AccountId>);
fn migrate_fee_storage() -> Self;
fn get_deposit_fee(&self) -> Option<Fee>;
fn get_withdraw_fee(&self) -> Option<Fee>;
fn get_deposit_fee_per_silo(&self, silo: AccountId) -> Option<Fee>;
fn get_withdraw_fee_per_silo(&self, silo: AccountId) -> Option<Fee>;
fn get_fee_owner(&self) -> AccountId;
```

Migration: 
The migration is needed due to the added field `fee` to the contract structure. To migrate the storage we need to call:
`pub fn migrate_fee_storage() -> Self`

## Performance / NEAR gas cost considerations
The gas for withdrawal and deposit transactions will be increased a little bit due to additional fee calculations and reads from the storage.

## Testing
Unit tests have been added to cover multiple fee cases.

## How should this be reviewed
The review should verify the functions `withdraw`, `engine_withdraw`, and `finish_deposit` and make sure that the burn/mint/transfer logic is correct.

### CODE REVIEW

File: eth-connector/src/lib.rs [hunk 1]
```diff
@@ -707,9 +892,182 @@ mod tests {
...
+            assert_eq!(
+                eth_connector_balance_after_deposit,
+                eth_connector_balance_before_deposit + test_case.expected_fee_amount,
+                "Wrong eth-connector balance after deposit"
+            );
+
+            owner_balance_before_deposit = owner_balance_after_deposit;
+            eth_connector_balance_before_deposit = eth_connector_balance_after_deposit;
+        }
+    }
```
Comment (olga24912) [inline eth-connector/src/lib.rs:1066]: We also need tests for silos and tests when we set `None` as fee.

**Review by olga24912 — CHANGES_REQUESTED** [review]

File: eth-connector/src/lib.rs [hunk 2]
```diff
@@ -525,8 +559,17 @@ impl EngineConnectorWithdraw for EthConnectorContract {
             .sdk_unwrap();
         // Burn tokens to recipient
         self.ft.internal_withdraw(&sender_id, amount);
+
+        let fee_amount = self.calculate_fee_amount(
+            amount.into(),
+            FeeType::Withdraw,
+            Some(env::predecessor_account_id()),
```
[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/lib.rs:566]: > Do we have different deployment of aurora-eth-connector for different engines? Or this contract should support all engines?

The purpose of that function is backward compatibility with the current Aurora Engine withdraw function. Newly deployed silos will not have such a function so will not need to access engine_withdraw.

> We should use the sender_id, not the env::predecessor_account_id(), the env::predecessor_account_id() is just a proxy, we decrement balance from sender_id

Yep, true

File: eth-connector/src/connector.rs [hunk 3]
```diff
@@ -10,6 +11,23 @@ pub trait Deposit {
...
+        &self,
+        amount: U128,
+        fee_type: FeeType,
+        silo: Option<AccountId>,
+    ) -> U128;
+    fn set_deposit_fee(&mut self, fee: Option<Fee>);
+    fn set_withdraw_fee(&mut self, fee: Option<Fee>);
+    fn set_deposit_fee_per_silo(&mut self, silo: AccountId, fee: Option<Fee>);
+    fn set_withdraw_fee_per_silo(&mut self, silo: AccountId, fee: Option<Fee>);
+    fn claim_fee(&mut self, amount: U128, receiver_id: Option<AccountId>);
```
[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/connector.rs:30]: 98bf6eedb2b710f2b9044e03cd52a035350d4dc2

File: eth-connector/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/lib.rs]: 26c186a32751d505b46f3fc80411c3f6dd8fa172

File: eth-connector/src/lib.rs (hunk 1 above)
Comment (karim-en) [reply eth-connector/src/lib.rs:1140]: c709963855d058e46d6d223fc2649cf7fde4156e

**Review by olga24912 — APPROVED** [review]

File: eth-connector/src/lib.rs [hunk 4]
```diff
@@ -540,6 +580,86 @@ impl Deposit for EthConnectorContract {
...
+        amount: U128,
+        fee_type: FeeType,
+        silo: Option<AccountId>,
+    ) -> U128 {
+        let Some(fee) = (match fee_type {
+            FeeType::Deposit => silo.and_then(|silo| self.fee.deposit_fee_per_silo.get(&silo)).or(self.fee.deposit_fee),
+            FeeType::Withdraw => silo.and_then(|silo| self.fee.withdraw_fee_per_silo.get(&silo)).or(self.fee.withdraw_fee),
+        }) else { return 0.into() };
+
+        let fee_amount = (amount.0 * fee.fee_percentage.0) / FEE_DECIMAL_PRECISION;
```
Comment (aleksuss) [inline eth-connector/src/lib.rs:612]: Does it make sense to use the `checked_mul` method instead?

**Review by aleksuss — APPROVED** [review]

File: eth-connector/src/lib.rs [hunk 5]
```diff
@@ -540,6 +580,87 @@ impl Deposit for EthConnectorContract {
...
+    }
+
+    fn claim_fee(&mut self, amount: U128, receiver_id: Option<AccountId>) {
+        self.connector.assert_owner_access_right().sdk_unwrap();
+        self.ft_transfer(
+            receiver_id.unwrap_or_else(env::predecessor_account_id),
+            amount,
+            None,
+        );
+    }
```
Comment (sept-en) [inline eth-connector/src/lib.rs:661]: I think it might be better to have a method that optionally allows you to collect fees on some different account. Thus, it won't need the owner access right of the ETH connector to collect the fees.

On the cons, it would require having an extra `ft_transfer()` call for each deposit/withdrawal and thus also consume more gas.

File: eth-connector/src/fee.rs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    Deposit,
+    Withdraw,
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Debug, Clone, Serialize, Deserialize, Copy)]
+pub struct Fee {
+    pub fee_percentage: U128,
+    pub lower_bound: Option<U128>,
+    pub upper_bound: Option<U128>,
+}
```
Comment (sept-en) [inline eth-connector/src/fee.rs:20]: Please describe all the fields.

File: eth-connector/src/fee.rs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+#[derive(Copy, Debug, Serialize, Deserialize, Clone)]
+pub enum FeeType {
+    Deposit,
+    Withdraw,
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Debug, Clone, Serialize, Deserialize, Copy)]
+pub struct Fee {
+    pub fee_percentage: U128,
```
Comment (sept-en) [inline eth-connector/src/fee.rs:17]: In this particular implementation, the fee is not a percentage but rather a coefficient of a numerator, while the denominator is equal to `1_000_000` due to 6-decimals precision. So, perhaps, the more proper naming would be `fee_numerator`? WDYT?

CC @karim-en @aleksuss @mrLSD

File: eth-connector/src/lib.rs [hunk 8]
```diff
@@ -556,31 +677,52 @@ impl FundsFinish for EthConnectorContract {
 
         log!("Finish deposit with the amount: {}", deposit_call.amount);
 
-        // Mint - calculate new balances
-        self.mint_eth_on_near(&deposit_call.new_owner_id, deposit_call.amount);
         // Store proof only after `mint` calculations
```
Comment (sept-en) [inline eth-connector/src/lib.rs:680]: This comment needs to be updated. 
Also, why don't we mint the whole amount to the `current_account_id` in the first place?

File: eth-connector/src/lib.rs [hunk 9]
```diff
@@ -498,8 +526,14 @@ impl Withdraw for EthConnectorContract {
         let sender_id = env::predecessor_account_id();
         // Burn tokens to recipient
         self.ft.internal_withdraw(&sender_id, amount);
+
+        let fee_amount =
```
Comment (sept-en) [inline eth-connector/src/lib.rs:530]: I think it's necessary to produce logs on the amount of fee and withdrawal amount. The same thing for the deposit.

File: eth-connector/src/lib.rs [hunk 10]
```diff
@@ -556,31 +677,52 @@ impl FundsFinish for EthConnectorContract {
...
+            );
+            match promise {
+                PromiseOrValue::Promise(p) => PromiseOrValue::Promise(p),
+                PromiseOrValue::Value(v) => PromiseOrValue::Value(Some(v)),
+            }
+        } else {
+            let fee_amount =
+                self.calculate_fee_amount(deposit_call.amount.into(), FeeType::Deposit, None);
+            let amount_to_transfer = deposit_call.amount.checked_sub(fee_amount.0).sdk_unwrap();
+
```
Comment (sept-en) [inline eth-connector/src/lib.rs:718]: In this case, it's rather `deposit_amount` as we don't create any transfer per se.

**Review by sept-en — CHANGES_REQUESTED** [review]

File: eth-connector/src/lib.rs [hunk 11]
```diff
@@ -525,8 +559,14 @@ impl EngineConnectorWithdraw for EthConnectorContract {
...
         self.ft.internal_withdraw(&sender_id, amount);
+
+        let fee_amount =
+            self.calculate_fee_amount(amount.into(), FeeType::Withdraw, Some(sender_id));
+        // Mint fee
+        self.mint_eth_on_near(&env::current_account_id(), fee_amount.0);
+
         WithdrawResult {
-            amount,
+            amount: amount.checked_sub(fee_amount.0).sdk_unwrap(),
```
[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/lib.rs:569]: It is dangerous to panic because other components/contracts that may call this method don't expect any panics and don't implement a refund functionality, so it was implemented as described in the specification:

> If the the calculated fee_amount is higher than the withdraw/deposit amount then charge the full amount as a fee and don't transfer anything to the recipient.

File: eth-connector/src/lib.rs [hunk 12]
```diff
@@ -498,8 +526,14 @@ impl Withdraw for EthConnectorContract {
...
```
[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/lib.rs:536]: It is dangerous to panic because other components/contracts that may call this method don't expect any panics and don't implement a refund functionality, so it was implemented as described in the specification:

If the the calculated fee_amount is higher than the withdraw/deposit amount then charge the full amount as a fee and don't transfer anything to the recipient.

File: eth-connector/src/fee.rs (hunk 7 above)
Comment (karim-en) [reply eth-connector/src/fee.rs:17]: PS: Actually the `precision` could be also part of this fee config instead of being hardcoded.

File: eth-connector/src/lib.rs (hunk 8 above)
Comment (karim-en) [reply eth-connector/src/lib.rs:680]: We mint to the `current_account_id` just in case there is an `ft-transfer-call` when depositing to eth account, but on deposit to near account, we can mint directly to the `new_owner_id`.
If are asking how this worked with one call, then it worked because the `new_owner_id` == `current_account_id` on this `ft-transfer-call`.
additional to that it is not clear why actually the `env::predecessor_account_id(),` was used instead of `env::current_account_id()`

File: eth-connector/src/lib.rs (hunk 10 above)
Comment (karim-en) [reply eth-connector/src/lib.rs]: 10d1ab3ac41e26e55b890e87629fee46333d37aa

File: eth-connector/src/lib.rs (hunk 8 above)
Comment (karim-en) [reply eth-connector/src/lib.rs]: 10d1ab3ac41e26e55b890e87629fee46333d37aa

File: eth-connector/src/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/lib.rs]: 10d1ab3ac41e26e55b890e87629fee46333d37aa

File: eth-connector/src/lib.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/lib.rs]: 10d1ab3ac41e26e55b890e87629fee46333d37aa

File: eth-connector/src/lib.rs [hunk 13]
```diff
@@ -188,6 +200,12 @@ impl EthConnectorContract {
             metadata: LazyOption::new(StorageKey::Metadata, Some(metadata)),
             used_proofs: LookupSet::new(StorageKey::Proof),
             known_engine_accounts: LookupSet::new(StorageKey::EngineAccounts),
+            fee: FeeStorage {
+                deposit_fee: None,
+                withdraw_fee: None,
+                withdraw_fee_per_silo: UnorderedMap::new(StorageKey::WithdrawFeePerSilo),
+                deposit_fee_per_silo: UnorderedMap::new(StorageKey::DespositFeePerSilo),
+            },
```
[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/lib.rs]: 10d1ab3ac41e26e55b890e87629fee46333d37aa

File: eth-connector/src/fee.rs [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    pub lower_bound: Option<U128>,
+    pub upper_bound: Option<U128>,
+}
+
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct FeeStorage {
+    pub deposit_fee: Option<Fee>,
+    pub withdraw_fee: Option<Fee>,
+    pub withdraw_fee_per_silo: UnorderedMap<AccountId, Fee>,
+    pub deposit_fee_per_silo: UnorderedMap<AccountId, Fee>,
```
[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/fee.rs]: 10d1ab3ac41e26e55b890e87629fee46333d37aa

File: eth-connector/src/lib.rs (hunk 9 above)
Comment (karim-en) [reply eth-connector/src/lib.rs]: 0e6d3639563e74e3ca908bab4e429e714da404a0

File: eth-connector/src/lib.rs [hunk 15]
```diff
@@ -556,31 +693,68 @@ impl FundsFinish for EthConnectorContract {
...
+            let amount_to_transfer = deposit_call.amount.checked_sub(fee_amount).sdk_unwrap();
+
+            near_sdk::log!(
+                "deposit_amount: {}, fee_amount: {}",
+                amount_to_transfer,
+                fee_amount
+            );
+
+            // Early return if the fee is higher than the transferred amount
+            if amount_to_transfer == 0 && fee_amount != 0 {
```
Comment (sept-en) [inline eth-connector/src/lib.rs:724]: Why is this a strange comparison (depending on the previous subtraction) instead of just a gt/lt comparison?

File: eth-connector/src/fee.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/fee.rs]: c4aec919755ea40eaee25e7601687a83cde40552

File: eth-connector/src/fee.rs (hunk 6 above)
Comment (karim-en) [reply eth-connector/src/fee.rs]: c4aec919755ea40eaee25e7601687a83cde40552

File: eth-connector/src/lib.rs (hunk 5 above)
Comment (karim-en) [reply eth-connector/src/lib.rs:690]: 136f9b106410f7bc784862b5e74e852f08fc59e7

File: eth-connector/src/fee.rs [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    // The minimum fee amount for transfer
+    pub lower_bound: Option<U128>,
+    // The maximum fee amount for transfer
+    pub upper_bound: Option<U128>,
+}
+
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct FeeStorage {
+    // Enables a default fee for deposit transfers
+    pub deposit_fee: Option<Fee>,
```
[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/fee.rs:28]: Actually good idea, I think it is possible to implement it in this way.
@sept-en @olga24912 WDYT?

[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/fee.rs:28]: As I see, the `AccoundId` doesn't implement `default()`, so it can be implemented in this way just by using `Option<AccountId>`

[the remark this answers is not in the document]
Comment (aleksuss) [reply eth-connector/src/fee.rs:28]: Our version of the `AccountId` derives the [Default](https://github.com/aurora-is-near/aurora-engine/blob/develop/engine-types/src/account_id.rs#L19) trait. But `Option<AccountId>` is a case as well.

[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/fee.rs]: 15fff9aa5d989b414dcb9388aafc3752f85726b5

File: eth-connector/src/lib.rs [hunk 17]
```diff
@@ -540,6 +595,83 @@ impl Deposit for EthConnectorContract {
...
+            .unwrap_or_else(env::current_account_id)
+    }
+
+    fn calculate_fee_amount(
+        &self,
+        amount: U128,
+        fee_type: FeeType,
+        silo: Option<AccountId>,
+    ) -> U128 {
+        let Some(fee) = (match fee_type {
```
Comment (aleksuss) [inline eth-connector/src/lib.rs:621]: Could it be replaced to:
```rust
let Some(fee) = (match fee_type {
    FeeType::Deposit => self.fee.deposit_fee_per_silo.get(&silo),
    FeeType::Withdraw => self.fee.withdraw_fee_per_silo.get(&silo),
}) else { return 0.into() };
```
?

Comment (karim-en) [reply eth-connector/src/lib.rs:621]: No, it cann't, because we need to get the the default fee if the fee for silo isn't set.
  It could be replaced with 
```
Deposit => self.fee.deposit_fee_per_silo.get(&silo).or_else(||self.fee.deposit_fee_per_silo.get(&None)),
Withdraw => self.fee.withdraw_fee_per_silo.get(&silo).or_else(||self.fee.withdraw_fee_per_silo.get(&None)),
```
But the problem of this is that if the `silo` is `None` then read storage will be called twice for the same key `None`.

Comment (aleksuss) [reply eth-connector/src/lib.rs:621]: >   because we need to get the default fee if the fee for silo isn't set

Got it.

File: eth-connector/src/fee.rs (hunk 7 above)
Comment (mrLSD) [reply eth-connector/src/fee.rs:18]: @karim-en it shouldn't be `U128`. Just `u32`.

File: eth-connector/src/lib.rs [hunk 18]
```diff
@@ -556,31 +688,68 @@ impl FundsFinish for EthConnectorContract {
...
+            let fee_amount = self
+                .calculate_fee_amount(
+                    deposit_call.amount.into(),
+                    FeeType::Deposit,
+                    Some(args.receiver_id.clone()),
+                )
+                .0;
+            let amount_to_transfer = deposit_call.amount.checked_sub(fee_amount).sdk_unwrap();
+
+            near_sdk::log!(
```
Comment (mrLSD) [inline eth-connector/src/lib.rs:709]: Use `crate::log!` instead.

File: eth-connector/src/lib.rs [hunk 19]
```diff
@@ -556,31 +688,68 @@ impl FundsFinish for EthConnectorContract {
...
+                PromiseOrValue::Promise(p) => PromiseOrValue::Promise(p),
+                PromiseOrValue::Value(v) => PromiseOrValue::Value(Some(v)),
+            }
+        } else {
+            let fee_amount = self
+                .calculate_fee_amount(deposit_call.amount.into(), FeeType::Deposit, None)
+                .0;
+            let deposit_amount = deposit_call.amount.checked_sub(fee_amount).sdk_unwrap();
+
+            near_sdk::log!(
```
Comment (mrLSD) [inline eth-connector/src/lib.rs:741]: Please use `crate::log!` instead.

**Review by mrLSD — CHANGES_REQUESTED** [review]

**Comment by mrLSD** [comment]
@karim-en please cover also with **integration tests**: `eth-connector-tests` crate.

**Comment by mrLSD** [comment]
@karim-en and I believe [AIP](https://github.com/aurora-is-near/AIPs/pull/5) also should be extended.

File: eth-connector/src/lib.rs [hunk 20]
```diff
@@ -525,8 +566,22 @@ impl EngineConnectorWithdraw for EthConnectorContract {
...
+
+        let fee_amount = self
+            .calculate_fee_amount(amount.into(), FeeType::Withdraw, Some(sender_id))
+            .0;
+        // Mint fee
+        self.mint_eth_on_near(&self.get_fee_owner(), fee_amount);
+
+        let withdraw_amount = amount.checked_sub(fee_amount).sdk_unwrap();
+
+        near_sdk::log!(
```
[the remark this answers is not in the document]
Comment (karim-en) [reply eth-connector/src/lib.rs]: 307d3a1fa2e02f8219d8003621066e3cdeaa2bb6

File: eth-connector/src/fee.rs (hunk 7 above)
Comment (karim-en) [reply eth-connector/src/fee.rs]: 9939668b661663eb9d830b3ab65ba2ea0596b6ed

**Review by mrLSD — APPROVED** [review]
